Performance: Classifier performance optimizations

Merinorus opened the pull request:
## Proposed change


This brings the classifier performance optimizations mentioned in this thread: https://github.com/paperless-ngx/paperless-ngx/discussions/10156.

Compared to the other PRs referencing this discussion, this one is probably the most complex (apologies for the reviewer!), but it also delivers the greatest performance improvements.

**Predictions are ~10x faster, and classifier training is ~30% faster.**

Time measurements were taken on my laptop using a 10-million-character document. These are rough estimates of the performance benefits and should be taken with a grain of salt.

I also added regression tests for the `preprocess_content` function to ensure that the classifier remains functionally unchanged.
This adds some overhead to the PR and the tests, but I found it helped me understand how the classifier works.

### Key modifications:
#### 1. Cache classifier vectorized content
Text preprocessing and vectorization consume a significant amount of CPU time, and the vectorized content is identical for all four matching steps. So, for each document, the first vectorized result is cached in Redis and reused for the next three steps. This divides the prediction time by ~3-4.
The cache expires after 5 minutes.
Example: for a 10M-character document, the cache used 8KB. Typical usage is 200–400B per document.



#### 2. Text preprocessing optimizations
- Initialize NLTK components (module, stop words, stemmer, etc.) only on the first call → saves up to ~3s.
- Reduce string copying and regex overhead → preprocessing time divided by ~5, with less garbage collection.
- Use an LRU cache for word stemming. The cache is shared across workers via Redis, except during classifier training (where it is counterproductive). → Stemming time divided by ~3 on long documents.
- Skip stemming for digits, as they are never stemmed anyway.

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

Merinorus committed:
Add regression test for classifier content preprocessing

Merinorus committed:
Classifier performance improvements

- Use a shared cache to avoid reprocessing the document content and its vectorized representation for each prediction (correspondents, tags, storage paths, and document type).
- Optimize text preprocessing with fewer string parsing and copy operations.
- Import advanced text processing modules only once.
- Use a shared cache for the stemmer function, except during classifier training to avoid unnecessary cache I/O.

Merinorus committed:
Add LRU Cache classes

Merinorus committed:
Classifier: modifications

- handle NLTK initialization AttributeError like before classifier optimization
- cover preprocess_content function
- add tests with advanced text processing enabled / disabled
- code simplification and cleaning
- changed redis key naming

Merinorus committed:
Use default DB read cache TTL for LRU cache TTL

stumpylog commented on src/documents/classifier.py:
```diff
@@ -442,10 +512,8 @@ def predict_document_type(self, content: str) -> int | None:
             return None
 
     def predict_tags(self, content: str) -> list[int]:
-        from sklearn.utils.multiclass import type_of_target
```
Imports like these were deliberate.  Not every path needed them, so it's a minor optimization to import only when needed

stumpylog commented on src/documents/classifier.py:
```diff
@@ -356,72 +384,114 @@ def content_generator() -> Iterator[str]:
+        shared_cache=True,
+    ) -> str:
+        """
+        Process to contents of a document, distilling it down into
```
-> `Process the contents`

stumpylog commented on src/documents/classifier.py:
```diff
@@ -356,72 +384,114 @@ def content_generator() -> Iterator[str]:
+            elif RE_DIGIT.search(word):
+                return word
+            else:
+                result = "" if word in self._stop_words else self._stemmer.stem(word)
```
Does returning an empty string make sense, rather than skipping?  The empty string is going to get stored and might be a really hot one in the cache

Merinorus replied:
Both work, but you're right. The stop word dictionary is small, but it makes more sense to not cache stop words in the "stemmer" cache. The code is clearer as well. Thanks!

Merinorus committed:
Classifier performance optimization: fixes after review

- Do not add stop words in the stemmer cache
- Put potentially heavy sklearn import back into the calling function
- Fix comment typo

Co-authored-by: Trenton H <797416+stumpylog@users.noreply.github.com>
